### PR TITLE
feat: per user and team feature flags

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

- Adds per-user and per-team feature flags to the database schema.
- Creates new tables `UserFeatures` and `TeamFeatures` to support feature flags for individual users and teams.
- Adds foreign key constraints to link these new tables with existing `Feature`, `users`, and `Team` tables.
- Updates the `User` and `Team` models in the Prisma schema to include a relation to `Feature`.
- Modifies the `Feature` model to include relations to `User` and `Team`.

### How to test?

1. Run the new migration to create the new tables and relationships.
2. Verify that the new tables `UserFeatures` and `TeamFeatures` are created in the database.
3. Check that you can assign features to specific users and teams through the API or admin interface.
4. Confirm that querying users and teams returns their associated features.
5. Ensure that deleting a user, team, or feature cascades properly in the new relationships.

### Why make this change?

This change allows for more granular control over feature availability. Instead of enabling features globally, we can now enable or disable features for specific users or teams. This flexibility supports better feature testing, gradual rollouts, and customized experiences for different user segments or team requirements.

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Ensure the database is up-to-date with the latest migrations.
- Create test users and teams in the system.
- Assign specific features to users and teams using the new relationships.
- Query users and teams to verify that their assigned features are correctly associated.
- Test the cascade delete functionality by removing users, teams, and features.
- Verify that the feature assignment process works correctly through the API or admin interface.

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

**Review by zomars — COMMENTED** [review]
As discussed with @keithwillcode and @emrysal shipping as a foundation for follow up use cases.

File: packages/prisma/schema.prisma [hunk 1]
```diff
@@ -436,6 +436,7 @@ model Team {
   activeOrgWorkflows     WorkflowsOnTeams[]
   attributes             Attribute[]
   smsLockReviewedByAdmin Boolean                 @default(false)
+  features               Feature[]               @relation("team_features")
```
Comment (keithwillcode) [inline packages/prisma/schema.prisma:439]: Why are we using snake casing on these?

File: packages/prisma/migrations/20240801210911_adds_per_user_and_team_feature_flags/migration.sql [hunk 2]
```diff
@@ -0,0 +1,35 @@
+-- CreateTable
+CREATE TABLE "_user_features" (
```
Comment (keithwillcode) [inline packages/prisma/migrations/20240801210911_adds_per_user_and_team_feature_flags/migration.sql:2]: These names are weird and not standard with our naming conventions. Why don't we make new models in the schema so we get better and more consistent naming? 
e.g. `UserFeature` with columns `Id`, `UserId`, `FeatureId`

**Review by keithwillcode — CHANGES_REQUESTED** [review]

File: packages/prisma/migrations/20240801210911_adds_per_user_and_team_feature_flags/migration.sql (hunk 2 above)
Comment (zomars) [reply packages/prisma/migrations/20240801210911_adds_per_user_and_team_feature_flags/migration.sql:2]: These are autogenerated from the prisma schema changes. IDK if customizing these would affect the type generator

File: packages/prisma/schema.prisma (hunk 1 above)
Comment (zomars) [reply packages/prisma/schema.prisma:439]: Named to be consistent with user_eventtypes that we already have

Comment (keithwillcode) [reply packages/prisma/schema.prisma:439]: Isn't the snake casing legacy and we should aim to have pascal casing moving forward?

File: packages/prisma/migrations/20240801210911_adds_per_user_and_team_feature_flags/migration.sql (hunk 2 above)
Comment (zomars) [reply packages/prisma/migrations/20240801210911_adds_per_user_and_team_feature_flags/migration.sql:2]: Tinkering into this

File: packages/prisma/schema.prisma (hunk 1 above)
Comment (zomars) [reply packages/prisma/schema.prisma]: adressed on latest changes

File: packages/prisma/schema.prisma [hunk 3]
```diff
@@ -1142,24 +1143,50 @@ model VerifiedEmail {
...
   @@index([stale])
 }
 
+model UserFeatures {
+  user       User     @relation(fields: [userId], references: [id], onDelete: Cascade)
+  userId     Int
+  feature    Feature  @relation(fields: [featureId], references: [slug], onDelete: Cascade)
+  featureId  String
+  assignedAt DateTime @default(now())
+  assignedBy String
```
Comment (keithwillcode) [inline packages/prisma/schema.prisma:1172]: Curious why you went with string here?

**Review by keithwillcode — APPROVED** [review]

File: packages/prisma/schema.prisma (hunk 3 above)
Comment (zomars) [reply packages/prisma/schema.prisma:1172]: Good point. Should be Int

Comment (zomars) [reply packages/prisma/schema.prisma:1172]: Will fix on a follow up since it auto merged

**Pull request merged**